File: src/editor/delta.ts
// Client half of the delta1 incremental-save protocol (brain#25c).
//
// The server half (pkg/flowgo/delta.go) accepts `X-Flowgo-Save:
// delta1` bodies carrying only what changed; this module decides WHAT
// changed. Two parts:
//
//   1. A dirty set fed by the mutation seam (mutations.ts). Every
//      mutator names a kind and the map it fired on, so the set is
//      keyed (map path → dirty kinds) plus a document-level flag for
//      mutations that can span maps (delete-with-submaps, default
//      shape) and an overflow flag for anything this module does not
//      recognise — overflow means "tracking is incomplete, send the
//      full document", never "guess".
//
//   2. buildDelta, a pure diff: given the exact body the server last
//      acknowledged and the live graph, it derives ops for the dirty
//      scopes ONLY. Deriving ops by diffing (rather than recording
//      ids at the mutation seam) is what makes deletes fall out
//      correctly for free: an id present in the base but gone from
//      the live graph is a delete, an id upserted-then-deleted inside
//      one debounce window never differs from the base at all, and a
//      mutation that arrives while a save is in flight simply still
//      differs from the base at the NEXT save. The dirty set is a
//      scope (which collections to compare), not a ledger — it must
//      be a superset of the real changes, and stale entries cost a
//      no-op compare, not a wrong op.
//
// The wire shapes here mirror pkg/flowgo/delta.go verbatim; that file
// is the source of truth.

export type DirtyKind = "box" | "text" | "line" | "stroke" | "image" | "edge";

const ID_KINDS = ["box", "text", "line", "stroke", "image"] as const;
const ALL_KINDS: readonly DirtyKind[] = [...ID_KINDS, "edge"];

// The graph field each id-carrying kind lives in.
const FIELD: Record<(typeof ID_KINDS)[number], string> = {
  box: "boxes",
  text: "texts",
  line: "lines",
  stroke: "strokes",
  image: "images",
};

export interface DirtySet {
  /** map path → kinds whose collection may differ from the base. */
  readonly maps: Map<string, Set<DirtyKind>>;
  /** A mutation that can touch any map / doc-level fields fired. */
  doc: boolean;
  /** A mutation this module doesn't understand fired → full save. */
  overflow: boolean;
}

const emptyDirty = (): DirtySet => ({
  maps: new Map(),
  doc: false,
  overflow: false,
});

let dirty: DirtySet = emptyDirty();

const kindsFor = (mapPath: string): Set<DirtyKind> => {
  let s = dirty.maps.get(mapPath);
  if (!s) {
    s = new Set();
    dirty.maps.set(mapPath, s);
  }
  return s;
};

/**
 * Record one mutation-seam event. Called by mutations.ts on EVERY
 * mutation, whether or not delta saves are active — the bookkeeping is
 * O(1) and keeping it unconditional means the set is already correct
 * the moment a capable server arms the delta path.
 */
export const markMutation = (kind: string, mapPath: string): void => {
  switch (kind) {
    case "box":
    case "text":
    case "line":
    case "stroke":
    case "image":
    case "edge":
      kindsFor(mapPath).add(kind);
      break;
    case "currentMap":
      // Spans kinds on one map (paste, align, palette sweep).
      for (const k of ALL_KINDS) kindsFor(mapPath).add(k);
      break;
    case "doc":
      // Spans maps (submap cascade on delete) and doc-level fields
      // (defaultShape) — the diff must look at everything.
      dirty.doc = true;
      break;
    default:
      // A mutation kind this module doesn't know cannot be scoped, so
      // tracking is no longer complete. Fail toward the full save.
      dirty.overflow = true;
  }
};

/** Read the current set without consuming it. */
export const peekDirty = (): DirtySet => dirty;

/**
 * Take ownership of the current set, leaving a fresh empty one so
 * mutations that land while a save is in flight accumulate separately
 * and survive to the next save.
 */
export const takeDirty = (): DirtySet => {
  const t = dirty;
  dirty = emptyDirty();
  return t;
};

/** Union a previously taken set back in (the save it scoped died). */
export const restoreDirty = (s: DirtySet): void => {
  for (const [path, kinds] of s.maps) {
    const cur = kindsFor(path);
    for (const k of kinds) cur.add(k);
  }
  if (s.doc) dirty.doc = true;
  if (s.overflow) dirty.overflow = true;
};

export const clearDirty = (): void => {
  dirty = emptyDirty();
};

// ---------------------------------------------------------------
// buildDelta — the pure diff.
// ---------------------------------------------------------------

// Structural views of the graph; elements stay opaque except for the
// id, exactly like the server's json.RawMessage handling.
interface DeltaMap {
  path: string;
  edges?: unknown[];
  [collection: string]: unknown;
}
export interface DeltaGraph {
  maps: DeltaMap[];
  defaultShape?: number;
}

interface DeltaOp {
  op: "upsert" | "delete" | "set-kind" | "set-edges" | "set-map" | "drop-map";
  kind?: string;
  map?: string;
  id?: string;
  item?: unknown;
  items?: unknown;
  edges?: unknown;
}

const idOf = (item: unknown): string | null => {
  if (typeof item !== "object" || item === null) return null;
  const id = (item as { id?: unknown }).id;
  return typeof id === "string" && id !== "" ? id : null;
};

// Diff one id-keyed collection into upsert/delete ops. Position is
// part of byte parity (upsertByID replaces in place, appends new), so
// after computing ops we verify the order they would produce server-
// side matches the live order; a mismatch (reorder) has no cheap op
// and collapses into one set-kind replacing the whole collection.
// Returns null only when an element has no usable id — that is lost
// tracking, and the caller falls back to the full save.
const diffKind = (
  kind: string,
  map: string,
  baseItems: unknown[],
  curItems: unknown[],
): DeltaOp[] | null => {
  const baseBy = new Map<string, unknown>();
  const baseIds: string[] = [];
  for (const it of baseItems) {
    const id = idOf(it);
    if (id === null) return null;
    baseBy.set(id, it);
    baseIds.push(id);
  }
  const ops: DeltaOp[] = [];
  const curIds = new Set<string>();
  const order: string[] = []; // live order, for the parity check
  const appended: string[] = [];
  let duplicate = false;
  for (const it of curItems) {
    const id = idOf(it);
    if (id === null) return null;
    if (curIds.has(id)) duplicate = true;
    curIds.add(id);
    order.push(id);
    const b = baseBy.get(id);
    if (b === undefined) {
      appended.push(id);
      ops.push({ op: "upsert", kind, map, item: it });
    } else if (JSON.stringify(it) !== JSON.stringify(b)) {
      ops.push({ op: "upsert", kind, map, item: it });
    }
  }
  for (const id of baseIds) {
    if (!curIds.has(id)) ops.push({ op: "delete", kind, map, id });
  }
  if (ops.length === 0) return ops;
  // Predicted server-side order: base order minus deletes, new ids
  // appended in op order.
  const predicted = baseIds.filter((id) => curIds.has(id)).concat(appended);
  const parity =
    !duplicate &&
    predicted.length === order.length &&
    predicted.every((id, i) => id === order[i]);
  if (!parity) {
    return [{ op: "set-kind", kind, map, items: curItems }];
  }
  return ops;
};

/**
 * Diff the live graph against the exact body the server acknowledged,
 * scoped by the dirty set, into a delta1 request body. Returns the
 * JSON string, or null when the change has no delta expression (lost
 * tracking, id-less elements, a removed map with surviving
 * descendants) — null always means "send the full document".
 */
export const buildDelta = (
  base: DeltaGraph,
  cur: DeltaGraph,
  scope: DirtySet,
  baseRevision: number,
): string | null => {
  if (scope.overflow) return null;
  const baseBy = new Map(base.maps.map((m) => [m.path, m]));
  const curBy = new Map(cur.maps.map((m) => [m.path, m]));
  const ops: DeltaOp[] = [];
  let doc: { defaultShape: number } | undefined;
  if (scope.doc) {
    // Map-set diff. drop-map takes the whole "/"-boundary subtree
    // (see dropMapAndSubtree in pkg/flowgo/delta.go), so: dedup
    // removed paths that already fall under a dropped ancestor, and
    // refuse the delta when a removed path still has a LIVE
    // descendant — dropping it would destroy a map the client kept.
    const removed = [...baseBy.keys()].filter((p) => !curBy.has(p)).sort();
    const drops: string[] = [];
    for (const p of removed) {
      if (drops.some((d) => p === d || p.startsWith(d + "/"))) continue;
      for (const q of curBy.keys()) {
        if (q.startsWith(p + "/")) return null;
      }
      drops.push(p);
    }
    for (const p of drops) ops.push({ op: "drop-map", map: p });
    for (const p of curBy.keys()) {
      if (!baseBy.has(p)) ops.push({ op: "set-map", map: p });
    }
    const shape = cur.defaultShape ?? 0;
    if (shape !== (base.defaultShape ?? 0)) doc = { defaultShape: shape };
  }
  // Content scope: doc-dirty widens to every live map and kind (a doc
  // mutation may have touched any of them); otherwise exactly the
  // dirty (map, kind) pairs.
  const scoped: Iterable<[string, Iterable<DirtyKind>]> = scope.doc
    ? [...curBy.keys()].map((p): [string, Iterable<DirtyKind>] => [p, ALL_KINDS])
    : scope.maps;
  for (const [path, kinds] of scoped) {
    const cm = curBy.get(path);
    if (!cm) {
      // Element dirt on a map that no longer exists: expressible only
      // when the doc diff above already dropped it.
      if (scope.doc) continue;
      return null;
    }
    const bm = baseBy.get(path);
    for (const kind of kinds) {
      if (kind === "edge") {
        // Edges carry no ids (see delta.go) — any difference replaces
        // the map's whole edge array.
        const be = (bm?.edges ?? []) as unknown[];
        const ce = (cm.edges ?? []) as unknown[];
        if (JSON.stringify(be) !== JSON.stringify(ce)) {
          ops.push({ op: "set-edges", map: path, edges: ce });
        }
        continue;
      }
      const field = FIELD[kind];
      const kindOps = diffKind(
        kind,
        path,
        (bm?.[field] ?? []) as unknown[],
        (cm[field] ?? []) as unknown[],
      );
      if (kindOps === null) return null;
      ops.push(...kindOps);
    }
  }
  const delta: { base: number; ops: DeltaOp[]; doc?: { defaultShape: number } } =
    { base: baseRevision, ops };
  if (doc) delta.doc = doc;
  return JSON.stringify(delta);
};

File: pkg/flowgo/delta.go
// Delta save protocol v1 (brain#25c) — the server half.
//
// A full /save decodes the ENTIRE document out of the request body —
// measured 85ms of a 152ms save at 100k boxes — to persist what is
// typically a one-box change. A delta body carries only the changed
// elements, so the O(document) JSON decode disappears from the save
// path. What deliberately stays O(document) is the write itself:
// whole-file serialize + fsync + atomic rename (47ms floor at 100k)
// is the accepted price of crash-safety and hand-editability, so the
// on-disk model does not change here at all.
//
// The contract that keeps a delta indistinguishable from a full save:
//
//   - Byte parity. Ops are applied to the parsed in-memory document
//     and the result goes through the SAME validate + serialize +
//     atomic-write path SaveLocalGraphFrom runs. There is exactly one
//     serializer; the delta path adds no second way to write bytes,
//     so the file a delta produces is byte-identical to a full save
//     of the equivalent end-state document.
//
//   - Revision guard. A delta only makes sense against the document
//     the client last saw. The base revision is checked under the
//     file mutex, immediately before apply, so a full save or MCP
//     write that slipped in between forces a 409 — and the client's
//     answer to 409 is a full save, which cannot conflict. The same
//     409 (never 400/500) covers "no parsed document to apply
//     against": whatever broke the cached parse, a full save fixes
//     it by replacing the file wholesale, so 409 steers the client
//     to exactly the request that will succeed.
//
//   - Atomicity. Ops apply in order to a deep COPY of the cached
//     graph; the copy is validated and persisted only when every op
//     applied cleanly. A delta that fails at op N leaves the cache,
//     the file, and the revision exactly as they were — the client
//     retries with a full save and loses nothing.
package flowgo

import (
	"encoding/json"
	"errors"
	"fmt"
	"strings"
)

// SaveModeHeader / SaveModeDelta1 are the protocol surface of the
// delta capability. /state advertises SaveModeDelta1 in a
// SaveModeHeader response header (next to X-Flowgo-Accept-Encoding —
// same shared-bundle reasoning: the editor must never send a delta to
// a server that didn't announce it). A /save request carrying
// `X-Flowgo-Save: delta1` has a Delta body; a /save without the
// header is a full document, exactly as before.
const (
	SaveModeHeader = "X-Flowgo-Save"
	SaveModeDelta1 = "delta1"
)

// ErrDeltaConflict maps to HTTP 409: the delta's base revision is not
// the document's current revision, or there is no parsed document to
// apply against. Both mean the same thing to the client — fall back
// to a full save.
var ErrDeltaConflict = errors.New("delta base conflicts with the current document")

// ErrDeltaInvalid maps to HTTP 400: the delta itself is malformed
// (unknown op, unknown kind, missing/undecodable payload field). The
// client treats any non-2xx as fall-back-to-full, but 400 tells it —
// and anyone reading logs — that retrying the same delta is useless.
var ErrDeltaInvalid = errors.New("invalid delta")

// Delta is the `X-Flowgo-Save: delta1` request body:
//
//	{"base": <revision>, "ops": [...], "doc": {"defaultShape": 1}?}
type Delta struct {
	// Base is the revision (X-Flowgo-Revision) the client applied its
	// local edits on top of. Guarded against Revision() at apply time.
	Base uint64    `json:"base"`
	Ops  []DeltaOp `json:"ops"`
	// Doc carries document-level fields. A pointer per field
	// distinguishes "not mentioned" from "set to the zero value"
	// (defaultShape 0 = rectangle is a real setting).
	Doc *DeltaDoc `json:"doc,omitempty"`
}

// DeltaDoc is the document-level part of a delta.
type DeltaDoc struct {
	DefaultShape *int `json:"defaultShape,omitempty"`
}

// DeltaOp is one operation. Which fields matter depends on Op; the
// payload fields stay raw JSON so each op decodes exactly the type it
// needs (the item schema is pkg/graph's element types, verbatim).
type DeltaOp struct {
	// Op is one of: upsert, delete, set-kind, set-edges, set-map,
	// drop-map.
	Op string `json:"op"`
	// Kind selects the element collection for upsert / delete /
	// set-kind: box, text, line, stroke, image. (Edges have no ids,
	// so they get their own whole-array op instead.)
	Kind string `json:"kind,omitempty"`
	// Map is the target map path. Required by every op.
	Map string `json:"map,omitempty"`
	// ID is the element to remove (delete only).
	ID string `json:"id,omitempty"`
	// Item is one full element (upsert), carrying its own id.
	Item json.RawMessage `json:"item,omitempty"`
	// Items is the full replacement collection (set-kind).
	Items json.RawMessage `json:"items,omitempty"`
	// Edges is the full replacement edge array (set-edges).
	Edges json.RawMessage `json:"edges,omitempty"`
}

// ApplyLocalDelta is ApplyLocalDeltaFrom with an anonymous writer.
func ApplyLocalDelta(d Delta) (uint64, error) {
	return ApplyLocalDeltaFrom(d, "")
}

// ApplyLocalDeltaFrom applies a delta1 payload to the local document
// and persists the result through the full-save write path. origin is
// the writer's identity for live-event echo suppression, exactly as
// in SaveLocalGraphFrom. On success it returns the revision the write
// produced (unchanged when the delta was a byte-level no-op — e.g.
// only deletes of already-missing ids — because the persist layer
// skips identical bytes and byte-identical content is nothing to
// notify subscribers about).
//
// Errors: ErrDeltaConflict → 409, ErrDeltaInvalid / ErrInvalidGraph →
// 400, anything else is a disk failure → 500.
func ApplyLocalDeltaFrom(d Delta, origin string) (uint64, error) {
	cfg.LocalFileMu.Lock()
	defer cfg.LocalFileMu.Unlock()
	// The guard runs under the file mutex, so "base == current" here
	// means no other writer can land between the check and our
	// persist — the revision cannot move until we release the lock.
	if cur := Revision(); d.Base != cur {
		return 0, fmt.Errorf("%w: delta base %d, document at revision %d", ErrDeltaConflict, d.Base, cur)
	}
	g, err := localGraphLocked()
	if err != nil {
		// Missing file, unreadable file, unparseable file: a delta has
		// nothing to apply against, but a full save would succeed by
		// replacing the file wholesale — so this is a 409 (retry as
		// full save), never a 400/500 the client would give up on.
		return 0, fmt.Errorf("%w: no document to apply against: %v", ErrDeltaConflict, err)
	}
	// Atomicity: ops mutate a deep copy. An op failing halfway leaves
	// the cached graph — and therefore disk — untouched, so a 400 is
	// always safe to answer with "nothing happened".
	work := cloneGraph(*g)
	if err := ApplyDeltaOps(&work, d); err != nil {
		return 0, err
	}
	// From here down this is the full-save path, step for step:
	// ValidateWritable gate (same reasons as SaveLocalGraphFrom — the
	// items came off the network, and one unwritable id would brick
	// the file for every later read), version stamp, one serializer,
	// atomic write.
	if errs := validateWritable(work); len(errs) > 0 {
		msgs := make([]string, len(errs))
		for i, e := range errs {
			msgs[i] = e.Error()
		}
		return 0, fmt.Errorf("%w: %s", ErrInvalidGraph, strings.Join(msgs, "; "))
	}
	work.Version = cfg.Version()
	if err := persistLocalBytesLocked([]byte(serialize(work)), origin); err != nil {
		return 0, err
	}
	// Keep the applied copy as the cached parse. SaveLocalGraphFrom
	// drops the cache because its payload is a whole document straight
	// off the network; this copy is a parsed graph plus structured
	// mutations — the same category as an MCP mutation, which the
	// cache contract in localfile.go already admits. Dropping it here
	// would put an O(document) re-parse back in front of the NEXT
	// delta, which is the cost this protocol exists to remove.
	localFile.hasGraph = true
	localFile.graph = work
	return Revision(), nil
}

// deltaKinds is the closed set of id-carrying element collections.
// Guarded up front so an unknown kind is a 400 even on ops that would
// otherwise no-op (delete on a missing map must still reject "kine":
// the client is confused, and silently swallowing that hides it).
func deltaKindKnown(kind string) bool {
	switch kind {
	case "box", "text", "line", "stroke", "image":
		return true
	}
	return false
}

// ApplyDeltaOps applies a delta's ops and doc-level fields to g in
// place, with no revision guard and no persistence — the pure half of
// ApplyLocalDeltaFrom, exported for hosts with their own storage (the
// website applies deltas against Postgres-stored graphs and must
// share these exact op semantics, not re-implement them). The caller
// owns atomicity: hand in a copy if a failed op must leave the
// original untouched.
func ApplyDeltaOps(g *Graph, d Delta) error {
	for i := range d.Ops {
		if err := applyDeltaOp(g, d.Ops[i]); err != nil {
			return fmt.Errorf("%w: ops[%d]: %v", ErrDeltaInvalid, i, err)
		}
	}
	if d.Doc != nil && d.Doc.DefaultShape != nil {
		g.DefaultShape = *d.Doc.DefaultShape
	}
	return nil
}

func applyDeltaOp(g *Graph, op DeltaOp) error {
	if op.Op != "" && op.Map == "" {
		return fmt.Errorf("op %q needs a map path", op.Op)
	}
	switch op.Op {
	case "upsert":
		if !deltaKindKnown(op.Kind) {
			return fmt.Errorf("unknown kind %q", op.Kind)
		}
		if len(op.Item) == 0 {
			return errors.New("upsert needs an item")
		}
		// Upsert implicitly materialises the map, mirroring the MCP
		// add_* contract — the editor may drop an element onto a
		// submap it just navigated into.
		return upsertInto(ensureMapAt(g, op.Map), op.Kind, op.Item)
	case "delete":
		if !deltaKindKnown(op.Kind) {
			return fmt.Errorf("unknown kind %q", op.Kind)
		}
		m := findMapAt(g, op.Map)
		if m == nil {
			// LWW tolerance: deleting from a map that is already gone
			// is as deleted as it gets. Same rule as a missing id.
			return nil
		}
		deleteFrom(m, op.Kind, op.ID)
		return nil
	case "set-kind":
		if !deltaKindKnown(op.Kind) {
			return fmt.Errorf("unknown kind %q", op.Kind)
		}
		if op.Items == nil {
			return errors.New("set-kind needs items")
		}
		return setKindOn(ensureMapAt(g, op.Map), op.Kind, op.Items)
	case "set-edges":
		if op.Edges == nil {
			return errors.New("set-edges needs edges")
		}
		var edges []Edge
		if err := json.Unmarshal(op.Edges, &edges); err != nil {
			return fmt.Errorf("bad edges: %v", err)
		}
		ensureMapAt(g, op.Map).Edges = edges
		return nil
	case "set-map":
		// Ensure-exists; an empty map serializes to nothing, so this
		// only changes bytes once something lands on the map.
		ensureMapAt(g, op.Map)
		return nil
	case "drop-map":
		dropMapAndSubtree(g, op.Map)
		return nil
	default:
		return fmt.Errorf("unknown op %q", op.Op)
	}
}

// dropMapAndSubtree removes the map at path and every map beneath it.
// The "/" boundary is the editor's withoutSubmaps rule
// (src/editor/factories.ts), ported exactly: a descendant is a path
// with the prefix path+"/", so dropping "/a" takes "/a/b" but must
// leave "/ab" alone — a bare prefix test would eat the sibling.
func dropMapAndSubtree(g *Graph, path string) {
	kept := g.Maps[:0]
	for _, m := range g.Maps {
		if m.Path == path || strings.HasPrefix(m.Path, path+"/") {
			continue
		}
		kept = append(kept, m)
	}
	g.Maps = kept
}

// upsertInto decodes item as the kind's element type and replaces the
// element with the same id, or appends when no such element exists.
func upsertInto(m *NamedMap, kind string, item json.RawMessage) error {
	decode := func(v any) error {
		if err := json.Unmarshal(item, v); err != nil {
			return fmt.Errorf("bad %s item: %v", kind, err)
		}
		return nil
	}
	switch kind {
	case "box":
		var v Box
		if err := decode(&v); err != nil {
			return err
		}
		m.Boxes = upsertByID(m.Boxes, func(b Box) string { return b.ID }, v)
	case "text":
		var v Text
		if err := decode(&v); err != nil {
			return err
		}
		m.Texts = upsertByID(m.Texts, func(t Text) string { return t.ID }, v)
	case "line":
		var v Line
		if err := decode(&v); err != nil {
			return err
		}
		m.Lines = upsertByID(m.Lines, func(l Line) string { return l.ID }, v)
	case "stroke":
		var v Stroke
		if err := decode(&v); err != nil {
			return err
		}
		m.Strokes = upsertByID(m.Strokes, func(s Stroke) string { return s.ID }, v)
	case "image":
		var v Image
		if err := decode(&v); err != nil {
			return err
		}
		m.Images = upsertByID(m.Images, func(i Image) string { return i.ID }, v)
	}
	return nil
}

// deleteFrom removes the element with the given id from the kind's
// collection. A missing id is a no-op by design (LWW tolerance): two
// clients deleting the same element must both get a 2xx, not a
// failure on whoever arrives second.
func deleteFrom(m *NamedMap, kind, id string) {
	switch kind {
	case "box":
		m.Boxes = deleteByID(m.Boxes, func(b Box) string { return b.ID }, id)
	case "text":
		m.Texts = deleteByID(m.Texts, func(t Text) string { return t.ID }, id)
	case "line":
		m.Lines = deleteByID(m.Lines, func(l Line) string { return l.ID }, id)
	case "stroke":
		m.Strokes = deleteByID(m.Strokes, func(s Stroke) string { return s.ID }, id)
	case "image":
		m.Images = deleteByID(m.Images, func(i Image) string { return i.ID }, id)
	}
}

// setKindOn replaces the map's whole collection of one kind — the
// bulk fallback for edits with no cheaper delta expression (multi-
// select drag, reorder).
func setKindOn(m *NamedMap, kind string, items json.RawMessage) error {
	decode := func(v any) error {
		if err := json.Unmarshal(items, v); err != nil {
			return fmt.Errorf("bad %s items: %v", kind, err)
		}
		return nil
	}
	switch kind {
	case "box":
		var v []Box
		if err := decode(&v); err != nil {
			return err
		}
		m.Boxes = v
	case "text":
		var v []Text
		if err := decode(&v); err != nil {
			return err
		}
		m.Texts = v
	case "line":
		var v []Line
		if err := decode(&v); err != nil {
			return err
		}
		m.Lines = v
	case "stroke":
		var v []Stroke
		if err := decode(&v); err != nil {
			return err
		}
		m.Strokes = v
	case "image":
		var v []Image
		if err := decode(&v); err != nil {
			return err
		}
		m.Images = v
	}
	return nil
}

// upsertByID replaces the element whose id matches item's, else
// appends. Replace-in-place preserves the element's position, which
// is part of byte parity: a moved box must serialize on the same line
// it always did.
func upsertByID[T any](s []T, id func(T) string, item T) []T {
	want := id(item)
	for i := range s {
		if id(s[i]) == want {
			s[i] = item
			return s
		}
	}
	return append(s, item)
}

// deleteByID removes the first element with the given id; absent ids
// leave the slice untouched.
func deleteByID[T any](s []T, id func(T) string, want string) []T {
	for i := range s {
		if id(s[i]) == want {
			return append(s[:i], s[i+1:]...)
		}
	}
	return s
}
